# Dataset Analysis

## CellTypist Dataset

In [2]:
import anndata as ad
import scanpy as sc
from skopt.space import Real, Categorical
import pandas as pd

# Load training data
adata = ad.io.read_h5ad('../data/CellTypistDataset/CountAdded_PIP_global_object_for_cellxgene_annotated_fine_grained.h5ad')

# Filter blood data
adata = adata[adata.obs['Organ'] == 'BLD'].copy()
print(adata)


# Preprocessing

# mitochondrial genes, "MT-" for human, "Mt-" for mouse
adata.var["mt"] = adata.var_names.str.startswith("MT-").to_numpy()
# ribosomal genes
adata.var["ribo"] = adata.var_names.str.startswith(("RPS", "RPL")).to_numpy()
# hemoglobin genes
adata.var["hb"] = adata.var_names.str.contains("^HB[^(P)]").to_numpy()

sc.pp.calculate_qc_metrics(adata, qc_vars=["mt", "ribo", "hb"], inplace=True, log1p=True)

# Remove mitochondrial, ribosomal and hemoglobin
adata = adata[:, ~adata.var["mt"]].copy()
adata = adata[:, ~adata.var["ribo"]].copy()
adata = adata[:, ~adata.var["hb"]].copy()

# Doublet Detection
sc.pp.scrublet(adata, batch_key="Donor")


# Normalization

# Saving count data
adata.layers["counts"] = adata.X.copy()

# Normalizing to median total counts
sc.pp.normalize_total(adata)
# Logarithmize the data
sc.pp.log1p(adata)


# Filtering Highly variable genes
#print('Before filtering highly variable genes ---')
#print(adata)

#sc.pp.highly_variable_genes(adata, n_top_genes=36473)

# Apply filter
#adata = adata[:, adata.var['highly_variable']].copy()

#print('After filtering highly variable genes ---')
#print(adata)

# Create train test split

# All Donors: ['621B', '637C', 'A35', 'A36', 'D496', 'D503']
donor_train = ['637C', 'A35', 'A36', 'D503']
donor_test = ['621B', 'D496']

adata_train = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()

adata_test = adata[
    adata.obs["Donor"].isin(donor_test)
].copy()

# Check split
print(adata_train.obs['Donor'].unique())
print(adata_test.obs['Donor'].unique())

# Prepare Data for training
X_train = adata_train.X#.to_df()
gene_names_train = adata_train.var_names
y_train = adata_train.obs['Manually_curated_celltype']

X_test = adata_test.X#.to_df()
gene_names_train = adata_train.var_names
y_test = adata_test.obs['Manually_curated_celltype']

AnnData object with n_obs × n_vars = 27620 × 36473
    obs: 'Organ', 'Donor', 'Chemistry', 'Cell_category', 'Predicted_labels_CellTypist', 'Majority_voting_CellTypist', 'Majority_voting_CellTypist_high', 'Manually_curated_celltype', 'Sex', 'Age_range', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', 'doublet_score', 'predicted_doublet', 'scumi-annotation'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Age_range_colors', 'Sex_colors', 'scrublet'
    obsm: 'X_umap'
    layers: 'counts'
['A36', 'A35', '637C', 'D503']
Categori

In [3]:
print("Train size: ", len(y_train))
print("Test size: ", len(y_test))

Train size:  18325
Test size:  9295


In [4]:
classes = adata.obs['Manually_curated_celltype']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

Manually_curated_celltype
Tnaive/CM_CD4               6609
Classical monocytes         5850
Teffector/EM_CD4            4331
NK_CD16+                    3539
Tem/emra_CD8                1983
Tnaive/CM_CD8               1182
Tregs                        798
Trm/em_CD8                   635
NK_CD56bright_CD16-          445
T_CD4/CD8                    270
Nonclassical monocytes       236
MNP/T doublets               229
Cycling T&NK                 215
Naive B cells                194
Memory B cells               171
Megakaryocytes               163
Tfh                          111
MAIT                         107
DC2                          100
Plasma cells                  66
T/B doublets                  62
Progenitor                    61
Tgd_CRTAM+                    47
pDC                           43
Trm_Th1/Th17                  35
Plasmablasts                  35
Mast cells                    22
ABCs                          21
ILC3                          15
DC1              

Cells remaining after filtering min_samples=150
Tnaive/CM_CD4               6609
Classical monocytes         5850
Teffector/EM_CD4            4331
NK_CD16+                    3539
Tem/emra_CD8                1983
Tnaive/CM_CD8               1182
Tregs                        798
Trm/em_CD8                   635
NK_CD56bright_CD16-          445
T_CD4/CD8                    270
Nonclassical monocytes       236
MNP/T doublets               229
Cycling T&NK                 215
Naive B cells                194
Memory B cells               171
Megakaryocytes               163

In [6]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("Number of classes present: ", len(class_counts))
print(class_counts)

Number of classes present:  20
scumi-annotation
CD14+ Monocyte              4032
Migratory dendritic cell    3985
CD8 TEM                     3200
CD56-bright NK cell         3080
CD4 TCM                     2853
CD4 CTL                     2201
Neutrophil                  2117
CD4 Naive T cell            2032
CD8 Naive T cell             928
CD56-dim NK cell             843
T regulatory cell            409
Naive B cell                 395
Pro-Neutrophil               374
Unassigned                   342
Gamma-delta T cell           337
Platelet                     151
Hematopoietic stem cell      131
Plasma cell                  106
Dark zone B cell              64
CD16+ Monocyte                40
Name: count, dtype: int64


Missing cells from marker genes:
CD16+ monocyte
Plasmacytoid dendritic cell

## Human Cell Atlas

In [ ]:
import anndata as ad
import scanpy as sc
from skopt.space import Real, Categorical
import pandas as pd

# Load training data
adata = ad.io.read_h5ad('../data/humancellatlas/5f29c29a-51c6-435c-8ff0-2b2a9d05ebee/BL_standard_design_annotated_fine_grained.h5ad')

print(adata)


# Preprocessing

# mitochondrial genes, "MT-" for human, "Mt-" for mouse
adata.var["mt"] = adata.var_names.str.startswith("MT-").to_numpy()
# ribosomal genes
adata.var["ribo"] = adata.var_names.str.startswith(("RPS", "RPL")).to_numpy()
# hemoglobin genes
adata.var["hb"] = adata.var_names.str.contains("^HB[^(P)]").to_numpy()

sc.pp.calculate_qc_metrics(adata, qc_vars=["mt", "ribo", "hb"], inplace=True, log1p=True)

# Remove mitochondrial, ribosomal and hemoglobin
adata = adata[:, ~adata.var["mt"]].copy()
adata = adata[:, ~adata.var["ribo"]].copy()
adata = adata[:, ~adata.var["hb"]].copy()

# Doublet Detection
sc.pp.scrublet(adata, batch_key="Donor")


# Normalization

# Saving count data
adata.layers["counts"] = adata.X.copy()

# Normalizing to median total counts
sc.pp.normalize_total(adata)
# Logarithmize the data
sc.pp.log1p(adata)


# Filtering Highly variable genes
#print('Before filtering highly variable genes ---')
#print(adata)

#sc.pp.highly_variable_genes(adata, n_top_genes=36473)

# Apply filter
#adata = adata[:, adata.var['highly_variable']].copy()

#print('After filtering highly variable genes ---')
print(adata)

AnnData object with n_obs × n_vars = 40000 × 25696
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'n_cells', 'percent_cells', 'robust', 'highly_variable_features', 'mean', 'var', 'hvf_loess', 'hvf_rank', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leide

In [8]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
CD4+ T cell                    6912
Cytotoxic T cell               2393
B cell                         1963
CD14+ monocyte                 1582
Natural killer cell            1335
CD16+ monocyte                  464
Dendritic cell                  252
Plasmacytoid dendritic cell      55
Megakaryocyte                    44
Name: count, dtype: int64


In [1]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/humancellatlas/5f29c29a-51c6-435c-8ff0-2b2a9d05ebee/BL_standard_design_annotated_fine_grained_more_donors.h5ad')

print(adata)


classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("Number of classes present: ", len(class_counts))
print(class_counts)

AnnData object with n_obs × n_vars = 40000 × 25696
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'n_cells', 'percent_cells', 'robust', 'highly_variable_features', 'mean', 'var', 'hvf_loess', 'hvf_rank', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leide

In [3]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/humancellatlas/5f29c29a-51c6-435c-8ff0-2b2a9d05ebee/BL_standard_design_annotated_fine_grained_more_donors.h5ad')

print(adata)


classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("Number of classes present: ", len(class_counts))
print(class_counts)

AnnData object with n_obs × n_vars = 39996 × 25696
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'n_cells', 'percent_cells', 'robust', 'highly_variable_features', 'mean', 'var', 'hvf_loess', 'hvf_rank', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leide

In [5]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/humancellatlas/5f29c29a-51c6-435c-8ff0-2b2a9d05ebee/BL_standard_design_annotated_fine_grained_more_donors.h5ad')

print(adata)

donor_train = ['4', '6', '7', '8']

adata = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()


classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("Number of classes present: ", len(class_counts))
print(class_counts)

AnnData object with n_obs × n_vars = 39996 × 25696
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'n_cells', 'percent_cells', 'robust', 'highly_variable_features', 'mean', 'var', 'hvf_loess', 'hvf_rank', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leide

In [1]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/humancellatlas/5f29c29a-51c6-435c-8ff0-2b2a9d05ebee/BL_standard_design_annotated_fine_grained_more_donors.h5ad')

print(adata)

donor_train = ['2', '3', '4', '6']
donor_test = ['7', '8']

adata_train = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()

adata_test = adata[
    adata.obs["Donor"].isin(donor_test)
].copy()


classes = adata_train.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TRAIN")
print("Number of classes present: ", len(class_counts))
print(class_counts)

classes = adata_test.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TEST")
print("Number of classes present: ", len(class_counts))
print(class_counts)

AnnData object with n_obs × n_vars = 39996 × 25696
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'n_cells', 'percent_cells', 'robust', 'highly_variable_features', 'mean', 'var', 'hvf_loess', 'hvf_rank', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leide

In [4]:
import itertools
import numpy as np
import pandas as pd


TARGET_MIN = 0.20
TARGET_MAX = 0.35
TARGET_CENTER = 0.275
DONOR_COL = "Donor"

donor_counts = adata.obs[DONOR_COL].value_counts().sort_index()
total_cells = donor_counts.sum()

print("Cells pro Donor:")
print(donor_counts)
print("\nGesamtzahl:", total_cells)

# Find best combinations
best_split = None
best_score = np.inf
donors = donor_counts.index.tolist()

# test all splits with exactly 2 donors for testing
for test_donors in itertools.combinations(donors, 2):
    test_cells = donor_counts[list(test_donors)].sum()
    test_ratio = test_cells / total_cells

    if test_ratio > TARGET_MAX or test_ratio < TARGET_MIN:
        continue

    # Score = distance to target middle (absolute value)
    score = abs(test_ratio - TARGET_CENTER)

    if score < best_score:
        train_donors = [
            d for d in donors if d not in test_donors
        ]
        best_score = score
        best_split = {
            "train_donors": train_donors,
            "test_donors": list(test_donors),
            "test_ratio": test_ratio,
            "test_cells": int(test_cells)
        }

print("\nBEST SPLIT")
print("-" * 40)

print("Train Donors:")
print(best_split["train_donors"])

print("\nTest Donors:")
print(best_split["test_donors"])

print(f"\nTest Anteil: {best_split['test_ratio']:.3f}")
print(f"Test Zellen: {best_split['test_cells']}")

# Apply split
train_adata = adata[
    adata.obs[DONOR_COL].isin(best_split["train_donors"])
].copy()

test_adata = adata[
    adata.obs[DONOR_COL].isin(best_split["test_donors"])
].copy()

# Validate best split
print("\nTrain shape:", train_adata.shape)
print("Test shape :", test_adata.shape)

Cells pro Donor:
Donor
2    6666
3    6666
4    6666
6    6666
7    6666
8    6666
Name: count, dtype: int64

Gesamtzahl: 39996

BEST SPLIT
----------------------------------------
Train Donors:
['4', '6', '7', '8']

Test Donors:
['2', '3']

Test Anteil: 0.333
Test Zellen: 13332

Train shape: (26664, 25696)
Test shape : (13332, 25696)


## Human Immune Health Atlas

In [18]:
import anndata as ad
import scanpy as sc
from skopt.space import Real, Categorical
import pandas as pd

# Load training data
adata_preprocessed = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated.h5ad')
adata = adata_preprocessed#.raw.to_adata()

# Filter blood data
#adata = adata[adata.obs['Organ'] == 'BLD'].copy()
print(adata)


# Preprocessing

# mitochondrial genes, "MT-" for human, "Mt-" for mouse
adata.var["mt"] = adata.var_names.str.startswith("MT-").to_numpy(dtype=bool)
# ribosomal genes
adata.var["ribo"] = adata.var_names.str.startswith(("RPS", "RPL")).to_numpy(dtype=bool)
# hemoglobin genes
adata.var["hb"] = adata.var_names.str.contains("^HB[^(P)]").to_numpy(dtype=bool)

sc.pp.calculate_qc_metrics(adata, qc_vars=["mt", "ribo", "hb"], inplace=True, log1p=True)

# Remove mitochondrial, ribosomal and hemoglobin
adata = adata[:, ~adata.var["mt"]].copy()
adata = adata[:, ~adata.var["ribo"]].copy()
adata = adata[:, ~adata.var["hb"]].copy()

# Doublet Detection
if "batch_id" not in adata.obs:
    print(
        "WARNUNG: 'batch_id' existiert nicht! Vorhandene Spalten:",
        adata.obs.columns.tolist(),
    )
# 1. Remove NaN values from batch_id after filtering
adata = adata[adata.obs["batch_id"].notna()].copy()

# 2. Filter empty cells and genes
sc.pp.filter_cells(adata, min_genes=1)
sc.pp.filter_genes(adata, min_cells=1)

# 3. Exclude too small batches
#batch_counts = adata.obs["batch_id"].value_counts()
#valid_batches = batch_counts[batch_counts >= 10].index
#adata = adata[adata.obs["batch_id"].isin(valid_batches)].copy()

# 3. Remove unused batch_ids
if str(adata.obs["batch_id"].dtype) == "category":
    adata.obs["batch_id"] = (
        adata.obs["batch_id"].cat.remove_unused_categories()
    )

# 4. Detect Doublets
#sc.pp.scrublet(adata, batch_key="Donor")
sc.pp.scrublet(adata, batch_key="batch_id")


# Normalization

# Saving count data
adata.layers["counts"] = adata.X.copy()

# Normalizing to median total counts
sc.pp.normalize_total(adata)
# Logarithmize the data
sc.pp.log1p(adata)


# Filtering Highly variable genes
print('Before filtering highly variable genes ---')
print(adata)

sc.pp.highly_variable_genes(adata, n_top_genes=36473)

# Apply filter
adata = adata[:, adata.var['highly_variable']].copy()

print('After filtering highly variable genes ---')
print(adata)

AnnData object with n_obs × n_vars = 31824 × 23921
    obs: 'cohort.cohortGuid', 'sample.sampleKitGuid', 'specimen.specimenGuid', 'pipeline.fileGuid', 'subject.subjectGuid', 'subject.biologicalSex', 'subject.birthYear', 'subject.ageAtFirstDraw', 'subject.ageGroup', 'subject.race', 'subject.ethnicity', 'subject.cmv', 'subject.bmi', 'sample.visitName', 'sample.drawYear', 'sample.subjectAgeAtDraw', 'batch_id', 'pool_id', 'chip_id', 'well_id', 'barcodes', 'original_barcodes', 'cell_name', 'n_reads', 'n_umis', 'n_genes', 'total_counts_mito', 'pct_counts_mito', 'doublet_score', 'AIFI_L1', 'AIFI_L2', 'AIFI_L3', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_co

In [ ]:
scumi-annotation
CD4+ T cell            13393
CD14+ monocyte          6481
Natural killer cell     4087
Cytotoxic T cell        2704
B cell                   395
Megakaryocyte            269
Dendritic cell           185
Plasma cell              106

In [20]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
Natural killer cell    11772
Cytotoxic T cell       10385
B cell                  3865
CD4+ T cell             2513
CD16+ monocyte          1885
CD14+ monocyte          1323
Megakaryocyte             81
Name: count, dtype: int64


In [21]:
classes = adata.obs['AIFI_L1']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

AIFI_L1
T cell             18531
Monocyte            6034
B cell              3257
NK cell             2512
DC                  1355
Platelet              79
Progenitor cell       22
Erythrocyte           20
ILC                   14
Name: count, dtype: int64


In [22]:
classes = adata.obs['AIFI_L2']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

AIFI_L2
Memory CD4 T cell        6326
CD14 monocyte            5161
Naive CD4 T cell         4350
Memory CD8 T cell        3369
CD56dim NK cell          2306
Naive CD8 T cell         1756
Naive B cell             1472
MAIT                     1276
Plasma cell               962
cDC2                      777
gdT                       659
Treg                      643
CD16 monocyte             556
Memory B cell             515
pDC                       496
Intermediate monocyte     317
Transitional B cell       178
CD56bright NK cell        152
Effector B cell           130
Proliferating T cell       79
Platelet                   79
Proliferating NK cell      54
cDC1                       49
DN T cell                  40
CD8aa                      33
ASDC                       33
Progenitor cell            22
Erythrocyte                20
ILC                        14
Name: count, dtype: int64


In [17]:
classes = adata.obs['AIFI_L3']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

AIFI_L3
Core CD14 monocyte         4563
Core naive CD4 T cell      3963
CM CD4 T cell              3696
GZMK- CD56dim NK cell      1901
Core naive CD8 T cell      1707
                           ... 
CLP cell                      7
ISG+ naive CD8 T cell         6
Activated memory B cell       4
KLRF1- effector Vd1 gdT       3
BaEoMaP cell                  1
Name: count, Length: 71, dtype: int64


In [4]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained.h5ad')

print(adata)


classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("Number of classes present: ", len(class_counts))
print(class_counts)

AnnData object with n_obs × n_vars = 81794 × 25324
    obs: 'cohort.cohortGuid', 'sample.sampleKitGuid', 'specimen.specimenGuid', 'pipeline.fileGuid', 'subject.subjectGuid', 'subject.biologicalSex', 'subject.birthYear', 'subject.ageAtFirstDraw', 'subject.ageGroup', 'subject.race', 'subject.ethnicity', 'subject.cmv', 'subject.bmi', 'sample.visitName', 'sample.drawYear', 'sample.subjectAgeAtDraw', 'batch_id', 'pool_id', 'chip_id', 'well_id', 'barcodes', 'original_barcodes', 'cell_name', 'n_reads', 'n_umis', 'n_genes', 'total_counts_mito', 'pct_counts_mito', 'doublet_score', 'AIFI_L1', 'AIFI_L2', 'AIFI_L3', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_co

## Comparison of available classes

In [ ]:
# CellTypist Dataset
CD14+ Monocyte              4032
Migratory dendritic cell    3985
CD8 TEM                     3200
CD56-bright NK cell         3080
CD4 TCM                     2853
CD4 CTL                     2201
Neutrophil                  2117
CD4 Naive T cell            2032
CD8 Naive T cell             928
CD56-dim NK cell             843
T regulatory cell            409
Naive B cell                 395
Pro-Neutrophil               374
Unassigned                   342
Gamma-delta T cell           337
Platelet                     151
Hematopoietic stem cell      131
Plasma cell                  106
Dark zone B cell              64
CD16+ Monocyte                40

In [ ]:
blood_immune_celltypes = [
    # T- & NK-Zellen
    'CD4 Naive T cell',  #CellTypist, Human Immune Health Atlas
    'CD4 TCM',  #CellTypist, Humman Cell Atlas
    'CD4 TEM', 
    'T regulatory cell',  #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
    'CD4 CTL',  #CellTypist
    'CD8 Naive T cell',  #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
    'CD8 TCM', # Human Immune Health Atlas
    'CD8 TEM',  #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
    'MAIT', # Humman Cell Atlas, Human Immune Health Atlas
    'Gamma-delta T cell',  #CellTypist, Human Immune Health Atlas
    'CD56-dim NK cell',  #CellTypist
    'CD56-bright NK cell', #CellTypist
    
    # B-Zellen & Plasmazellen
    'Naive B cell',  #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
    'Memory B cell', #Humman Cell Atlas, Human Immune Health Atlas
    'Plasma cell', #CellTypist, Human Immune Health Atlas
    
    # Myeloide Zellen & Dendritische Zellen
    'CD14+ Monocyte', #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
    'CD16+ Monocyte',  #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
    'CD1C+ dendritic cell', # Humman Cell Atlas, Human Immune Health Atlas
    'CLEC9A+ dendritic cell', 
    'Plasmacytoid dendritic cell', # Humman Cell Atlas, Human Immune Health Atlas
    
    # Granulozyten
    'Neutrophil', 
    'Basophil'
]
['T follicular helper cell', 
 'Pro B cell', 
 'Pre B cell', 
 'Dark zone B cell', #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
 'Light zone B cell',
 'Migratory dendritic cell', #CellTypist
 'Follicular dendritic cell',
 'Hematopoietic stem cell', #CellTypist, Human Immune Health Atlas
 'Erythroid cell',
 'Platelet', #CellTypist, Humman Cell Atlas, Human Immune Health Atlas
 'Pro-Neutrophil', #CellTypist
 'Pre-Neutrophil',
 'M1 macrophage',
 'M2 macrophage',
 'Mast cell']

Unassigned  #CellTypist

In [ ]:
# Humman Cell Atlas
CD8 Naive T cell               8346
CD4 TCM                        6754
CD14+ Monocyte                 4346
CD8 TCM                        4268
Gamma-delta T cell             3671
CD8 TEM                        3467
Naive B cell                   3024
Memory B cell                  2297
CD16+ Monocyte                 1177
MAIT                           1100
CD1C+ dendritic cell            675
T regulatory cell               512
Platelet                        122
Dark zone B cell                121
Plasmacytoid dendritic cell     120

In [ ]:
# Human Immune Health Atlas
CD4 Naive T cell               16513
CD8 TCM                        14690
CD8 TEM                        13267
CD14+ Monocyte                 12982
CD8 Naive T cell                6660
Naive B cell                    3791
MAIT                            3137
CD16+ Monocyte                  1976
Gamma-delta T cell              1817
Plasma cell                     1677
T regulatory cell               1633
CD1C+ dendritic cell            1366
Memory B cell                    987
Plasmacytoid dendritic cell      776
Platelet                         234
Dark zone B cell                 218
Hematopoietic stem cell           70

## Merged Dataset

In [15]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/CellTypist_HumanCellAtlas_Merged.h5ad')

print(adata)

donor_train = ['637C', 'A35', 'A36', 'D503', '2', '3', '4', '6']
donor_test = ['621B', 'D496', '7', '8']

adata_train = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()

adata_test = adata[
    adata.obs["Donor"].isin(donor_test)
].copy()

adata_ood = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained.h5ad')


import pandas as pd
import scanpy as sc

def compare_two_anndatas(
    adata1, 
    adata2, 
    cell_type_col='scumi-annotation', 
    name1='Train', 
    name2='Test'
):
    """
    Vergleicht die Zelltypen und deren Frequenzen zwischen zwei AnnData-Objekten.
    """
    # 1. Häufigkeiten der Zelltypen ermitteln
    counts1 = adata1.obs[cell_type_col].value_counts()
    counts2 = adata2.obs[cell_type_col].value_counts()

    # 2. In einem DataFrame zusammenführen
    df = pd.DataFrame({
        f'{name1}_Count': counts1, 
        f'{name2}_Count': counts2
    }).fillna(0).astype(int)

    # 3. Berechnungen durchführen
    df['Total_Count'] = df[f'{name1}_Count'] + df[f'{name2}_Count']
    df[f'{name1}_Ratio_%'] = (df[f'{name1}_Count'] / df['Total_Count'] * 100).round(1)
    df[f'{name2}_Ratio_%'] = (df[f'{name2}_Count'] / df['Total_Count'] * 100).round(1)

    # 4. Status bestimmen (Exklusivität)
    df['Status'] = 'In beiden'
    df.loc[df[f'{name1}_Count'] == 0, 'Status'] = f'Nur {name2}'
    df.loc[df[f'{name2}_Count'] == 0, 'Status'] = f'Nur {name1}'

    df = df.sort_values(by='Total_Count', ascending=False)

    # 5. Zusammenfassung in der Konsole ausgeben
    only_in_1 = df[df['Status'] == f'Nur {name1}'].index.tolist()
    only_in_2 = df[df['Status'] == f'Nur {name2}'].index.tolist()

    print("=" * 65)
    print(f"ZELLTYP-VERGLEICH: {name1} vs. {name2}")
    print("=" * 65)
    print(f"Gesamtanzahl unterschiedlicher Zelltypen : {len(df)}")
    print(f"In beiden Datensätzen vorhanden          : {len(df[df['Status'] == 'In beiden'])}")
    print(f"Ausschließlich in '{name1}' ({len(only_in_1):2d})            : {only_in_1 if only_in_1 else 'Keine'}")
    print(f"Ausschließlich in '{name2}' ({len(only_in_2):2d})            : {only_in_2 if only_in_2 else 'Keine'}")
    print("=" * 65 + "\n")

    return df

# ==========================================
# BEISPIEL-AUFRUF:
# ==========================================
df_comparison = compare_two_anndatas(
    adata_train, 
    adata_test, 
    name1='Merged_Train', 
    name2='Merged_Test'
)

df_comparison = compare_two_anndatas(
    adata_train, 
    adata_ood, 
    name1='Merged_Train', 
    name2='OOD_Test'
)

# Vollständige Tabelle anzeigen:
# pd.set_option('display.max_rows', None)
# print(df_comparison)

AnnData object with n_obs × n_vars = 67274 × 24668
    obs: 'Donor', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', 'doublet_score', 'predicted_doublet', 'scumi-annotation', 'dataset'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'featureid'
    obsm: 'X_umap'
ZELLTYP-VERGLEICH: Merged_Train vs. Merged_Test
Gesamtanzahl unterschiedlicher Zelltypen : 26
In beiden Datensätzen vorhanden          : 23
Ausschließlich in 'Merged_Train' ( 3)            : ['Migratory dendritic cell', 'Neutrophil', 'Dark zone B cell']
Ausschließlich in 'M

In [16]:
import pandas as pd

# 1. Maske für wahre NK-Zellen laut Autoren-Annotation (AIFI_L2) erstellen
nk_mask = adata_ood.obs['AIFI_L1'] == 'NK cell'

# Falls in AIFI_L2 Subtypen wie 'CD56dim NK cell' stehen, nutze alternativ:
# nk_mask = adata_ood.obs['AIFI_L2'].str.contains('NK', na=False)

# 2. Vorhergesagte Klassen aus 'scumi-annotation' für diese wahren NK-Zellen auswerten
nk_predictions = adata_ood.obs.loc[nk_mask, 'scumi-annotation'].value_counts()

# 3. Ergebnis und Fehlklassifikations-Statistik ausgeben
print("===========================================================")
print("Vorhergesagte Zelltypen ('scumi-annotation') für echte NK-Zellen:")
print("===========================================================")
print(nk_predictions)

total_nk = nk_mask.sum()
correct_nk = nk_predictions.get('NK cell', 0)
misclassified = total_nk - correct_nk

print("\n===========================================================")
print(f"Gesamtanzahl echter NK-Zellen: {total_nk}")
print(f"Richtig erkannt als 'NK cell': {correct_nk} ({correct_nk / total_nk * 100:.1f}%)")
print(f"Falsch klassifiziert:          {misclassified} ({misclassified / total_nk * 100:.1f}%)")
print("===========================================================")

Vorhergesagte Zelltypen ('scumi-annotation') für echte NK-Zellen:
scumi-annotation
CD8 TEM                        5894
Dark zone B cell                 68
Gamma-delta T cell                4
CD1C+ dendritic cell              0
CD4 Naive T cell                  0
CD8 TCM                           0
CD8 Naive T cell                  0
CD16+ Monocyte                    0
CD14+ Monocyte                    0
Hematopoietic stem cell           0
MAIT                              0
Memory B cell                     0
Naive B cell                      0
Plasma cell                       0
Plasmacytoid dendritic cell       0
Platelet                          0
T regulatory cell                 0
Name: count, dtype: int64

Gesamtanzahl echter NK-Zellen: 5966
Richtig erkannt als 'NK cell': 0 (0.0%)
Falsch klassifiziert:          5966 (100.0%)


In [6]:
classes = adata_train.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TRAIN")
print("Number of classes present: ", len(class_counts))
print(class_counts)

classes = adata_test.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TEST")
print("Number of classes present: ", len(class_counts))
print(class_counts)

TRAIN
Number of classes present:  26
scumi-annotation
CD4 Naive T cell               6325
CD14+ Monocyte                 5709
CD8 TCM                        4837
Migratory dendritic cell       3985
CD8 TEM                        3388
CD4 TCM                        2849
Naive B cell                   2672
CD4 CTL                        2200
CD56-dim NK cell               2194
CD8 Naive T cell               2171
Neutrophil                     2117
Memory B cell                  1477
MAIT                           1375
CD16+ Monocyte                  788
Gamma-delta T cell              520
T regulatory cell               399
Pro-Neutrophil                  354
CD1C+ dendritic cell            289
T follicular helper cell        230
Plasma cell                     185
Platelet                        174
CD56-bright NK cell             171
Plasmacytoid dendritic cell     131
Hematopoietic stem cell         129
Dark zone B cell                 64
CLEC9A+ dendritic cell           28
Name: coun

In [17]:
classes = adata_ood.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TEST")
print("Number of classes present: ", len(class_counts))
print(class_counts)

TEST
Number of classes present:  17
scumi-annotation
CD4 Naive T cell               16513
CD8 TCM                        14690
CD8 TEM                        13267
CD14+ Monocyte                 12982
CD8 Naive T cell                6660
Naive B cell                    3791
MAIT                            3137
CD16+ Monocyte                  1976
Gamma-delta T cell              1817
Plasma cell                     1677
T regulatory cell               1633
CD1C+ dendritic cell            1366
Memory B cell                    987
Plasmacytoid dendritic cell      776
Platelet                         234
Dark zone B cell                 218
Hematopoietic stem cell           70
Name: count, dtype: int64


In [11]:
classes = adata_ood.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TEST")
print("Number of classes present: ", len(class_counts))
print(class_counts)

TEST
Number of classes present:  17
scumi-annotation
CD4 Naive T cell               15596
CD8 TCM                        13753
CD8 TEM                        13253
CD14+ Monocyte                 11541
CD8 Naive T cell                6200
Naive B cell                    5164
Gamma-delta T cell              3493
MAIT                            3065
CD16+ Monocyte                  1759
T regulatory cell               1540
CD1C+ dendritic cell            1127
Memory B cell                   1036
Plasmacytoid dendritic cell      784
Platelet                         220
Dark zone B cell                 198
Hematopoietic stem cell           57
Unassigned                        46
Name: count, dtype: int64


In [ ]:
Delete: 'Migratory dendritic cell', both Neutrophils, CLEC9A+ dendritic cell
Merge: NK Cells

Probleme: gar keine NK Cells im OOD Dataset

In [3]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/CellTypist_HumanCellAtlas_Merged.h5ad')

print(adata)

donor_train = ['637C', 'A35', 'A36', 'D503', '2', '3', '4', '6']
donor_test = ['621B', 'D496', '7', '8']

adata_train = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()

adata_test = adata[
    adata.obs["Donor"].isin(donor_test)
].copy()

adata_ood = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained.h5ad')


import pandas as pd
import scanpy as sc

def compare_two_anndatas(
    adata1, 
    adata2, 
    cell_type_col='scumi-annotation', 
    name1='Train', 
    name2='Test'
):
    """
    Vergleicht die Zelltypen und deren Frequenzen zwischen zwei AnnData-Objekten.
    """
    # 1. Häufigkeiten der Zelltypen ermitteln
    counts1 = adata1.obs[cell_type_col].value_counts()
    counts2 = adata2.obs[cell_type_col].value_counts()

    # 2. In einem DataFrame zusammenführen
    df = pd.DataFrame({
        f'{name1}_Count': counts1, 
        f'{name2}_Count': counts2
    }).fillna(0).astype(int)

    # 3. Berechnungen durchführen
    df['Total_Count'] = df[f'{name1}_Count'] + df[f'{name2}_Count']
    df[f'{name1}_Ratio_%'] = (df[f'{name1}_Count'] / df['Total_Count'] * 100).round(1)
    df[f'{name2}_Ratio_%'] = (df[f'{name2}_Count'] / df['Total_Count'] * 100).round(1)

    # 4. Status bestimmen (Exklusivität)
    df['Status'] = 'In beiden'
    df.loc[df[f'{name1}_Count'] == 0, 'Status'] = f'Nur {name2}'
    df.loc[df[f'{name2}_Count'] == 0, 'Status'] = f'Nur {name1}'

    df = df.sort_values(by='Total_Count', ascending=False)

    # 5. Zusammenfassung in der Konsole ausgeben
    only_in_1 = df[df['Status'] == f'Nur {name1}'].index.tolist()
    only_in_2 = df[df['Status'] == f'Nur {name2}'].index.tolist()

    print("=" * 65)
    print(f"ZELLTYP-VERGLEICH: {name1} vs. {name2}")
    print("=" * 65)
    print(f"Gesamtanzahl unterschiedlicher Zelltypen : {len(df)}")
    print(f"In beiden Datensätzen vorhanden          : {len(df[df['Status'] == 'In beiden'])}")
    print(f"Ausschließlich in '{name1}' ({len(only_in_1):2d})            : {only_in_1 if only_in_1 else 'Keine'}")
    print(f"Ausschließlich in '{name2}' ({len(only_in_2):2d})            : {only_in_2 if only_in_2 else 'Keine'}")
    print("=" * 65 + "\n")

    return df

# ==========================================
# BEISPIEL-AUFRUF:
# ==========================================
df_comparison = compare_two_anndatas(
    adata_train, 
    adata_test, 
    name1='Merged_Train', 
    name2='Merged_Test'
)

df_comparison = compare_two_anndatas(
    adata_train, 
    adata_ood, 
    name1='Merged_Train', 
    name2='OOD_Test'
)

# Vollständige Tabelle anzeigen:
# pd.set_option('display.max_rows', None)
# print(df_comparison)

AnnData object with n_obs × n_vars = 67274 × 24668
    obs: 'Donor', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', 'doublet_score', 'predicted_doublet', 'scumi-annotation', 'dataset'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'featureid'
    obsm: 'X_umap'
ZELLTYP-VERGLEICH: Merged_Train vs. Merged_Test
Gesamtanzahl unterschiedlicher Zelltypen : 26
In beiden Datensätzen vorhanden          : 23
Ausschließlich in 'Merged_Train' ( 3)            : ['Migratory dendritic cell', 'Neutrophil', 'Dark zone B cell']
Ausschließlich in 'M

In [12]:
print(adata_ood)

AnnData object with n_obs × n_vars = 78832 × 25281
    obs: 'cohort.cohortGuid', 'sample.sampleKitGuid', 'specimen.specimenGuid', 'pipeline.fileGuid', 'subject.subjectGuid', 'subject.biologicalSex', 'subject.birthYear', 'subject.ageAtFirstDraw', 'subject.ageGroup', 'subject.race', 'subject.ethnicity', 'subject.cmv', 'subject.bmi', 'sample.visitName', 'sample.drawYear', 'sample.subjectAgeAtDraw', 'batch_id', 'pool_id', 'chip_id', 'well_id', 'barcodes', 'original_barcodes', 'cell_name', 'n_reads', 'n_umis', 'n_genes', 'total_counts_mito', 'pct_counts_mito', 'doublet_score', 'AIFI_L1', 'AIFI_L2', 'AIFI_L3', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_co

In [5]:
import pandas as pd

# 1. Maske für wahre NK-Zellen laut Autoren-Annotation (AIFI_L2) erstellen
nk_mask = adata_ood.obs['AIFI_L2'] == 'Memory CD4 T cell' #'NK cell'

# Falls in AIFI_L2 Subtypen wie 'CD56dim NK cell' stehen, nutze alternativ:
# nk_mask = adata_ood.obs['AIFI_L2'].str.contains('NK', na=False)

# 2. Vorhergesagte Klassen aus 'scumi-annotation' für diese wahren NK-Zellen auswerten
nk_predictions = adata_ood.obs.loc[nk_mask, 'scumi-annotation'].value_counts()

# 3. Ergebnis und Fehlklassifikations-Statistik ausgeben
print("===========================================================")
print("Vorhergesagte Zelltypen ('scumi-annotation') für echte NK-Zellen:")
print("===========================================================")
print(nk_predictions)

total_nk = nk_mask.sum()
correct_nk = nk_predictions.get('NK cell', 0)
misclassified = total_nk - correct_nk

print("\n===========================================================")
print(f"Gesamtanzahl echter NK-Zellen: {total_nk}")
print(f"Richtig erkannt als 'NK cell': {correct_nk} ({correct_nk / total_nk * 100:.1f}%)")
print(f"Falsch klassifiziert:          {misclassified} ({misclassified / total_nk * 100:.1f}%)")
print("===========================================================")

Vorhergesagte Zelltypen ('scumi-annotation') für echte NK-Zellen:
scumi-annotation
CD8 TCM                        12737
CD4 Naive T cell                 745
CD8 TEM                          410
MAIT                             170
T regulatory cell                139
CD8 Naive T cell                  33
Gamma-delta T cell                 6
Dark zone B cell                   1
CD1C+ dendritic cell               0
CD14+ Monocyte                     0
CD16+ Monocyte                     0
Memory B cell                      0
Hematopoietic stem cell            0
Naive B cell                       0
Plasma cell                        0
Plasmacytoid dendritic cell        0
Platelet                           0
Name: count, dtype: int64

Gesamtanzahl echter NK-Zellen: 14241
Richtig erkannt als 'NK cell': 0 (0.0%)
Falsch klassifiziert:          14241 (100.0%)


In [6]:
classes = adata_ood.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TEST")
print("Number of classes present: ", len(class_counts))
print(class_counts)

TEST
Number of classes present:  17
scumi-annotation
CD4 Naive T cell               16513
CD8 TCM                        14690
CD8 TEM                        13267
CD14+ Monocyte                 12982
CD8 Naive T cell                6660
Naive B cell                    3791
MAIT                            3137
CD16+ Monocyte                  1976
Gamma-delta T cell              1817
Plasma cell                     1677
T regulatory cell               1633
CD1C+ dendritic cell            1366
Memory B cell                    987
Plasmacytoid dendritic cell      776
Platelet                         234
Dark zone B cell                 218
Hematopoietic stem cell           70
Name: count, dtype: int64


In [11]:
import pandas as pd
import anndata as ad

adata_ood_clean = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained_cleaned.h5ad')

scumi_col = 'scumi_clean'
author_col = 'AIFI_L2'

# Alle eindeutigen Klassen aus scumi_clean durchlaufen
unique_classes = sorted(adata_ood_clean.obs[scumi_col].dropna().unique())

print(f"STARTE EVALUIERUNG: {len(unique_classes)} Klassen in '{scumi_col}' vs. '{author_col}'\n")

for cell_type in unique_classes:
    # 1. Maske für die aktuelle Klasse in scumi_clean erstellen
    cell_mask = adata_ood_clean.obs[scumi_col] == cell_type
    total_cells = cell_mask.sum()
    
    # 2. Häufigkeiten der Autoren-Labels für diese Zellen ermitteln
    author_distribution = adata_ood_clean.obs.loc[cell_mask, author_col].value_counts()
    
    # 3. Strukturierte Ausgabe
    print("=" * 65)
    print(f"Klasse in '{scumi_col}': {cell_type} (Gesamt: {total_cells} Zellen)")
    print("=" * 65)
    
    if total_cells == 0:
        print("  [Keine Zellen gefunden]\n")
        continue

    for author_label, count in author_distribution.items():
        percentage = (count / total_cells) * 100
        print(f"  - {author_label:<35}: {count:>6} ({percentage:>5.1f}%)")
    print()

STARTE EVALUIERUNG: 15 Klassen in 'scumi_clean' vs. 'AIFI_L2'

Klasse in 'scumi_clean': CD14+ Monocyte (Gesamt: 12318 Zellen)
  - CD14 monocyte                      :  12318 (100.0%)
  - CD16 monocyte                      :      0 (  0.0%)
  - CD56bright NK cell                 :      0 (  0.0%)
  - CD56dim NK cell                    :      0 (  0.0%)
  - MAIT                               :      0 (  0.0%)
  - Memory B cell                      :      0 (  0.0%)
  - Memory CD4 T cell                  :      0 (  0.0%)
  - Memory CD8 T cell                  :      0 (  0.0%)
  - Naive B cell                       :      0 (  0.0%)
  - Naive CD4 T cell                   :      0 (  0.0%)
  - Naive CD8 T cell                   :      0 (  0.0%)
  - Plasma cell                        :      0 (  0.0%)
  - Proliferating NK cell              :      0 (  0.0%)
  - Treg                               :      0 (  0.0%)
  - cDC2                               :      0 (  0.0%)
  - gdT            

In [7]:
import anndata as ad
import pandas as pd

adata = ad.io.read_h5ad('../data/CellTypist_HumanCellAtlas_Merged.h5ad')

print(adata)

donor_train = ['D496', 'D503', '2', '3', '4', '6']
donor_test = ['621B', '637C', 'A35', 'A36', '7', '8']

adata_train = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()

adata_test = adata[
    adata.obs["Donor"].isin(donor_test)
].copy()

adata_ood = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained.h5ad')


import pandas as pd
import scanpy as sc

def compare_two_anndatas(
    adata1, 
    adata2, 
    cell_type_col='scumi-annotation', 
    name1='Train', 
    name2='Test'
):
    """
    Vergleicht die Zelltypen und deren Frequenzen zwischen zwei AnnData-Objekten.
    """
    # 1. Häufigkeiten der Zelltypen ermitteln
    counts1 = adata1.obs[cell_type_col].value_counts()
    counts2 = adata2.obs[cell_type_col].value_counts()

    # 2. In einem DataFrame zusammenführen
    df = pd.DataFrame({
        f'{name1}_Count': counts1, 
        f'{name2}_Count': counts2
    }).fillna(0).astype(int)

    # 3. Berechnungen durchführen
    df['Total_Count'] = df[f'{name1}_Count'] + df[f'{name2}_Count']
    df[f'{name1}_Ratio_%'] = (df[f'{name1}_Count'] / df['Total_Count'] * 100).round(1)
    df[f'{name2}_Ratio_%'] = (df[f'{name2}_Count'] / df['Total_Count'] * 100).round(1)

    # 4. Status bestimmen (Exklusivität)
    df['Status'] = 'In beiden'
    df.loc[df[f'{name1}_Count'] == 0, 'Status'] = f'Nur {name2}'
    df.loc[df[f'{name2}_Count'] == 0, 'Status'] = f'Nur {name1}'

    df = df.sort_values(by='Total_Count', ascending=False)

    # 5. Zusammenfassung in der Konsole ausgeben
    only_in_1 = df[df['Status'] == f'Nur {name1}'].index.tolist()
    only_in_2 = df[df['Status'] == f'Nur {name2}'].index.tolist()

    print("=" * 65)
    print(f"ZELLTYP-VERGLEICH: {name1} vs. {name2}")
    print("=" * 65)
    print(f"Gesamtanzahl unterschiedlicher Zelltypen : {len(df)}")
    print(f"In beiden Datensätzen vorhanden          : {len(df[df['Status'] == 'In beiden'])}")
    print(f"Ausschließlich in '{name1}' ({len(only_in_1):2d})            : {only_in_1 if only_in_1 else 'Keine'}")
    print(f"Ausschließlich in '{name2}' ({len(only_in_2):2d})            : {only_in_2 if only_in_2 else 'Keine'}")
    print("=" * 65 + "\n")

    return df

# ==========================================
# BEISPIEL-AUFRUF:
# ==========================================
df_comparison = compare_two_anndatas(
    adata_train, 
    adata_test, 
    name1='Merged_Train', 
    name2='Merged_Test'
)

df_comparison = compare_two_anndatas(
    adata_train, 
    adata_ood, 
    name1='Merged_Train', 
    name2='OOD_Test'
)

# Vollständige Tabelle anzeigen:
# pd.set_option('display.max_rows', None)
# print(df_comparison)

AnnData object with n_obs × n_vars = 67274 × 24668
    obs: 'Donor', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', 'doublet_score', 'predicted_doublet', 'scumi-annotation', 'dataset'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'featureid'
    obsm: 'X_umap'
ZELLTYP-VERGLEICH: Merged_Train vs. Merged_Test
Gesamtanzahl unterschiedlicher Zelltypen : 26
In beiden Datensätzen vorhanden          : 23
Ausschließlich in 'Merged_Train' ( 3)            : ['Migratory dendritic cell', 'CD4 TCM', 'T regulatory cell']
Ausschließlich in 'Mer

In [8]:
classes = adata_train.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TRAIN")
print("Number of classes present: ", len(class_counts))
print(class_counts)

classes = adata_test.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print("TEST")
print("Number of classes present: ", len(class_counts))
print(class_counts)

TRAIN
Number of classes present:  26
scumi-annotation
CD14+ Monocyte                 6795
CD8 TCM                        4837
CD4 Naive T cell               4361
Migratory dendritic cell       3985
CD8 TEM                        3881
CD8 Naive T cell               2979
CD56-dim NK cell               2966
CD56-bright NK cell            2911
CD4 TCM                        2853
Naive B cell                   2561
CD4 CTL                        2199
Neutrophil                     2113
Memory B cell                  1477
MAIT                           1375
CD16+ Monocyte                  788
Gamma-delta T cell              522
T regulatory cell               409
Pro-Neutrophil                  363
CD1C+ dendritic cell            289
T follicular helper cell        230
Plasma cell                     197
Plasmacytoid dendritic cell     131
Hematopoietic stem cell         105
Dark zone B cell                 60
Platelet                         43
CLEC9A+ dendritic cell           28
Name: coun

### Clean Annotations

#### OOD Dataset: Human Immune Health Atlas

In [7]:
import scanpy as sc
import anndata as ad
import pandas as pd

TARGET_CLASSES = [
    'CD4 Naive T cell',
    'CD4 Memory T cell',
    'CD14+ Monocyte',
    'CD8 Naive T cell',
    'NK cell',
    'Naive B cell',
    'MAIT',
    'CD16+ Monocyte',
    'Gamma-delta T cell',
    'Plasma cell',
    'T regulatory cell',
    'CD1C+ dendritic cell',
    'Memory B cell',
    'Plasmacytoid dendritic cell',
    'CD8 Memory T cell'
]

merge_dict = {
    'CD56-dim NK cell': 'NK cell',
    'CD56-bright NK cell': 'NK cell',
    'CD4 TCM': 'CD4 Memory T cell',
    'CD4 CTL': 'CD4 Memory T cell',
    'T follicular helper cell': 'CD4 Memory T cell',
    'CD8 TEM': 'CD8 Memory T cell',
    'CD8 TCM': 'CD8 Memory T cell',
    'Dark zone B cell': 'Memory B cell'
}

aifi_l2_to_target_map = {
    # B-Zellen & Plasmazellen
    'Plasma cell': 'Plasma cell',
    'Memory B cell': 'Memory B cell',
    'Effector B cell': 'Effector B cell',
    'Naive B cell': 'Naive B cell',
    'Transitional B cell': 'Transitional B cell',
    
    # T-Zellen
    'Naive CD4 T cell': 'CD4 Naive T cell',
    'Memory CD4 T cell': 'CD4 Memory T cell',
    'CD4 TEM': 'CD4 Memory T cell',
    'CD4 TCM': 'CD4 Memory T cell',
    
    'Naive CD8 T cell': 'CD8 Naive T cell',
    'Memory CD8 T cell': 'CD8 Memory T cell',
    'CD8 TEM': 'CD8 Memory T cell',
    'CD8 TCM': 'CD8 Memory T cell',
    'CD8aa': 'CD8aa',
    
    'Treg': 'T regulatory cell',
    'MAIT': 'MAIT',
    'gdT': 'Gamma-delta T cell',
    
    # Myeloische / Dendritische / NK-Zellen
    'cDC2': 'CD1C+ dendritic cell',
    'CD1C+ dendritic cell': 'CD1C+ dendritic cell',
    'pDC': 'Plasmacytoid dendritic cell',
    'CD14 monocyte': 'CD14+ Monocyte',
    'CD16 monocyte': 'CD16+ Monocyte',
    'CD56bright NK cell': 'NK cell',
    'CD56dim NK cell': 'NK cell',

    'ASDC': 'ASDC',
    'Progenitor cell': 'Progenitor cell',
    'Intermediate monocyte': 'Intermediate monocyte',
}

remove_list = [
    'Migratory dendritic cell',
    'Neutrophil',
    'Pro-Neutrophil',
    'Platelet',
    'Hematopoietic stem cell',
    'CLEC9A+ dendritic cell'
]

def clean_and_harmonize_ood(adata, annotation_col='scumi-annotation', author_col='AIFI_L1'):
    adata_clean = adata.copy()
    
    # 1. String-Konvertierung für Manipulation
    adata_clean.obs[annotation_col] = adata_clean.obs[annotation_col].astype(str)
    adata_clean.obs[author_col] = adata_clean.obs[author_col].astype(str)
    adata_clean.obs['AIFI_L2'] = adata_clean.obs['AIFI_L2'].astype(str)
    
    # 2. Erstes Standard-Merging anwenden
    adata_clean.obs['scumi_clean'] = adata_clean.obs[annotation_col].replace(merge_dict)
    
    # 3.1 ZWINGENDES ÜBERSCHREIBEN: Alle Zellen, die laut Autoren (AIFI_L1) NK-Zellen sind, 
    #    werden in 'scumi_clean' auf 'NK cell' gesetzt
    author_nk_mask = adata_clean.obs[author_col].str.contains('NK', na=False)
    adata_clean.obs.loc[author_nk_mask, 'scumi_clean'] = 'NK cell'

    # 3.2 Alle Zellen die laut Autoren Memory CD4 T Zellen sind, werden in 'scumi_clean'
    #     auf 'CD4 Memory T cell' gesetzt
    author_cd4_memory_mask = adata_clean.obs['AIFI_L2'].str.contains('Memory CD4 T cell', na=False)
    adata_clean.obs.loc[author_cd4_memory_mask, 'scumi_clean'] = 'CD4 Memory T cell'

    # 3.3 GEZIELTES AUFLÖSEN DER PROBLEMATISCHEN KLASSEN ÜBER AIFI_L2
    problematic_classes = ['CD8 TCM', 'CD8 TEM', 'CD8 Memory T cell', 'Memory B cell', 'Plasma cell', 
                           'CD4 Naive T cell', 'CD8 Naive T cell', 'MAIT', 'Naive B cell', 'T regulatory cell',
                          'Plasmacytoid dendritic cell', 'Gamma-delta T cell', 'CD1C+ dendritic cell',
                           'CD16+ Monocyte', 'CD14+ Monocyte']
    prob_mask = adata_clean.obs['scumi_clean'].isin(problematic_classes)
    
    # Mappe die AIFI_L2 Autoren-Labels auf deine 15 Ziel-Klassen
    resolved_labels = adata_clean.obs.loc[prob_mask, 'AIFI_L2'].map(aifi_l2_to_target_map)
    
    # Überschreibe die problematischen Stellen
    adata_clean.obs.loc[prob_mask, 'scumi_clean'] = resolved_labels
    
    # 4. Filtern: Unerwünschte / seltene Zelltypen aus dem Dataset entfernen
    #keep_mask = ~adata_clean.obs['scumi_clean'].isin(remove_list)
    #adata_clean = adata_clean[keep_mask].copy()
    # 4. STRIKTE FILTERUNG: Behalte AUSSCHLIESSLICH deine 15 Ziel-Klassen
    keep_mask = adata_clean.obs['scumi_clean'].isin(TARGET_CLASSES)
    adata_clean = adata_clean[keep_mask].copy()
    
    # 5. Kategorien zurücksetzen
    adata_clean.obs['scumi_clean'] = adata_clean.obs['scumi_clean'].astype('category')
    adata_clean.obs[author_col] = adata_clean.obs[author_col].astype('category')
    adata_clean.obs['AIFI_L2'] = adata_clean.obs['AIFI_L2'].astype('category')
    
    # Kontrollausgabe
    print(f"Zellen nach Filterung: {adata_clean.n_obs}")
    print(f"NK-Zellen in scumi_clean nach Überschreiben: {(adata_clean.obs['scumi_clean'] == 'NK cell').sum()}")
    print(f"CD4 Memory T-Zellen in scumi_clean: {(adata_clean.obs['scumi_clean'] == 'CD4 Memory T cell').sum()}")
    print(f"CD8 Memory T-Zellen in scumi_clean: {(adata_clean.obs['scumi_clean'] == 'CD8 Memory T cell').sum()}")
    print("\nVerteilung aller 15 Ziel-Klassen:")
    print(adata_clean.obs['scumi_clean'].value_counts())
    
    return adata_clean

# Ausführen
adata_ood = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained.h5ad')
adata_ood_clean = clean_and_harmonize_ood(adata_ood, annotation_col='scumi-annotation', author_col='AIFI_L1')

Zellen nach Filterung: 79332
NK-Zellen in scumi_clean nach Überschreiben: 5966
CD4 Memory T-Zellen in scumi_clean: 14241
CD8 Memory T-Zellen in scumi_clean: 6665

Verteilung aller 15 Ziel-Klassen:
scumi_clean
CD4 Naive T cell               17161
CD4 Memory T cell              14241
CD14+ Monocyte                 12318
CD8 Memory T cell               6665
CD8 Naive T cell                6517
NK cell                         5966
Naive B cell                    3386
MAIT                            2891
Gamma-delta T cell              2274
CD16+ Monocyte                  1903
T regulatory cell               1731
Memory B cell                   1395
CD1C+ dendritic cell            1173
Plasma cell                      986
Plasmacytoid dendritic cell      725
Name: count, dtype: int64


In [10]:
ad.settings.allow_write_nullable_strings = True

# Save merged dataset
adata_ood_clean.write(filename='../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated_fine_grained_cleaned.h5ad')

#### ID Dataset: Merged CellTypist and Human Cell Atlas

In [2]:
import scanpy as sc
import pandas as pd
import anndata as ad

# 1. Definieren der Mappings (Merging) und zu entfernenden Zelltypen (Filtering)
merge_dict = {
    # NK-Zell Subtypen -> NK cell
    'CD56-dim NK cell': 'NK cell',
    'CD56-bright NK cell': 'NK cell',
    
    # CD4 Gedächtnis- und Spezial-Zellen -> CD4 Memory T cell
    'CD4 TCM': 'CD4 Memory T cell',
    'CD4 CTL': 'CD4 Memory T cell',
    'T follicular helper cell': 'CD4 Memory T cell',

    # CD8 Memory T cell
    'CD8 TEM': 'CD8 Memory T cell',
    'CD8 TCM': 'CD8 Memory T cell',
    
    # Keimzentrum B-Zellen -> Memory B cell
    'Dark zone B cell': 'Memory B cell'
}

remove_list = [
    'Migratory dendritic cell',
    'Neutrophil',
    'Pro-Neutrophil',
    'Platelet',
    'Hematopoietic stem cell',
    'CLEC9A+ dendritic cell'
]

def clean_and_harmonize_ood(adata, annotation_col='scumi-annotation', author_col='AIFI_L1'):
    """
    Führt Merging und Filtering auf dem OOD-AnnData Objekt durch.
    """
    adata_clean = adata.copy()
    
    # String-Konvertierung für saubere Bearbeitung (falls Categorical)
    adata_clean.obs[annotation_col] = adata_clean.obs[annotation_col].astype(str)
    
    # 2. Merging auf den Marker-basierten Labels anwenden
    adata_clean.obs['scumi_clean'] = adata_clean.obs[annotation_col].replace(merge_dict)
    
    # (Optional) Falls Autoren-Labels AIFI_L1 ebenfalls gemappt/harmonisierbar sind
    if author_col in adata_clean.obs.columns:
        adata_clean.obs[author_col] = adata_clean.obs[author_col].astype(str)
        adata_clean.obs['author_clean'] = adata_clean.obs[author_col].replace(merge_dict)
    
    # 3. Filtern: Entferne unerwünschte / seltene Zelltypen aus dem Dataset
    keep_mask = ~adata_clean.obs['scumi_clean'].isin(remove_list)
    adata_clean = adata_clean[keep_mask].copy()
    
    # Kategorische Typen zurücksetzen
    adata_clean.obs['scumi_clean'] = adata_clean.obs['scumi_clean'].astype('category')
    if author_col in adata_clean.obs.columns:
        adata_clean.obs['author_clean'] = adata_clean.obs['author_clean'].astype('category')
        
    print(f"Ursprüngliche Zellanzahl (OOD): {adata.n_obs}")
    print(f"Bereinigte Zellanzahl (OOD):   {adata_clean.n_obs}")
    print(f"Entfernte Zellen:              {adata.n_obs - adata_clean.n_obs}\n")
    print("Verteilung der bereinigten Klassen (scumi_clean):")
    print(adata_clean.obs['scumi_clean'].value_counts())
    
    return adata_clean

# Anwendung auf adata_ood
adata_merged = ad.io.read_h5ad('../data/CellTypist_HumanCellAtlas_Merged.h5ad')
adata_merged_clean = clean_and_harmonize_ood(adata_merged, annotation_col='scumi-annotation', author_col='AIFI_L1')

Ursprüngliche Zellanzahl (OOD): 67274
Bereinigte Zellanzahl (OOD):   60447
Entfernte Zellen:              6827

Verteilung der bereinigten Klassen (scumi_clean):
scumi_clean
CD8 Memory T cell              12258
CD4 Naive T cell                9580
CD14+ Monocyte                  9170
NK cell                         7772
CD4 Memory T cell               5527
Naive B cell                    4854
CD8 Naive T cell                4079
Memory B cell                   2487
MAIT                            1894
CD16+ Monocyte                   971
Gamma-delta T cell               639
T regulatory cell                409
CD1C+ dendritic cell             405
Plasma cell                      239
Plasmacytoid dendritic cell      163
Name: count, dtype: int64


In [3]:
ad.settings.allow_write_nullable_strings = True

# Save merged dataset
adata_merged_clean.write(filename='../data/CellTypist_HumanCellAtlas_Merged_cleaned.h5ad')